sdk/core/Azure.Core/tests/OperationInternalTests.cs

```diff
@@ -87,6 +87,67 @@ public void RawResponseInitialization()
             }
         }
 
+        [Test]
+        public void SetStateSucceeds()
+        {
+            var operationInternal = CreateOperation(isOfT, UpdateResult.Pending);
+            if (operationInternal is OperationInternal oi)
+            {
+                oi.SetState(OperationState.Success(mockResponse));
+            }
+            else if (operationInternal is OperationInternal<int> oit)
+            {
+                oit.SetState(OperationState<int>.Success(mockResponse, 1));
+            }
+
+            Assert.IsTrue(operationInternal.HasCompleted);
+            if (operationInternal is OperationInternal<int> oit2)
+            {
+                Assert.IsTrue(oit2.HasValue);
+                Assert.AreEqual(1, oit2.Value);
+            }
+        }
+
+        [Test]
+        public void SetStateIsPending()
+        {
+            var operationInternal = CreateOperation(isOfT, UpdateResult.Pending);
+            if (operationInternal is OperationInternal oi)
+            {
+                oi.SetState(OperationState.Pending(mockResponse));
+            }
+            else if (operationInternal is OperationInternal<int> oit)
+            {
+                oit.SetState(OperationState<int>.Pending(mockResponse));
+            }
+
+            Assert.IsFalse(operationInternal.HasCompleted);
+            if (operationInternal is OperationInternal<int> oit2)
+            {
+                Assert.IsFalse(oit2.HasValue);
+            }
+        }
+
+        [Test]
+        public void SetStateFails()
+        {
+            var operationInternal = CreateOperation(isOfT, UpdateResult.Pending);
+            if (operationInternal is OperationInternal oi)
+            {
+                Assert.Throws<RequestFailedException>(() => oi.SetState(OperationState.Failure(mockResponse)));
```

Comment:
This is an interesting behavior. Should we be failing inline in whatever place called SetState?

Reply:
`SetState` calls `ApplyToStateAsync(false, ...)` which throws on failure. Would we want to change that behavior? I could add another optional bool to throw or not, but what is the use case here? Are we sure another call to `UpdateState/Async` would occur?

Reply:
But, yes, I agree this was unexpected in a call to `SetState` but figured it was probably rare to call it and even more rare to set it here with failure. I would expect most people using `OperationInternal` would be returning state via `UpdateState(OperationState, CancellationToken)` (or generic).

Reply:
I think throwing inline is the side effect of `ApplyToStateAsync` being only used for `UpdateStatus` implementation. It feels plausible that someone might want to return a pre-failed Operation (Imagine they got the status in the first response). Can you please add a parameter to the ApplyToStateAsync and set it to false for the direct SetState call?

Reply:
I'll add the param, but as we discussed offline, I'll open a bug to consider not throwing for `UpdateStatus` at all which, IIRC, is how LROs were supposed to behavior previously. That means we need to throw somewhere, and if there's no `Value` it's not clear where. That's a whole other issue, though. The separate param will solve the unexpected exception with `SetState` here.